### ISSUE TITLE
Ability to select primary category for posts.

### ISSUE THREAD

**Issue by sybrew** [sybrew/the-seo-framework#32]
From: [Levente's question on WordPress](https://wordpress.org/support/topic/primary-category/).

When multiple categories are assigned to a post, the primary category is automatically chosen first of all set categories in alphabetic order.

The goal is to be able to select the primary category, as per the wish of the post author or editor.

Pros:
- LD+json and theme/plugin breadcrumbs will correctly display the preferred/primary category.
- Complex breadcrumbs will no longer be prioritized.
- The user has a more reliable expectation.

Cons:
- A possible cause for bugs with some themes and plugins.
- Requires additional interaction for each post.
- Could confuse or irritate users as they don't know what's set from just "viewing" the post in the overview or on the front-end.
- ~~Creates additional overhead in maintenance (WordPress 4.4+ differs greatly from what's below)~~ We support WP 4.4+ only now.

**Comment by tonybahama** [comment]
Hi!

Thanks for the help with the feature request. Actually, this would (as far as Yoast is concerned) only affect the product URL if you use the "category" slug in the URL.

User confusion can be easily solved by a tooltip on hover saying this would be in the permalink and that is it.

It does not need interaction, the default is to set the category as you set it now, automatically. It only gives an **ability** to change it.

About the bugs, I can't really add value here unfortunately :)

Anyways, I hope this gets implemented as our URL structure is important for seo reasons and selecting a main category helps with that.

**Comment by sybrew** [comment]
> It does not need interaction, the default is to set the category as you set it now, automatically. It only gives an ability to change it.

That's a great point.

To elaborate:
1. Option is default hidden.
2. When hovering over the category selection, the "make primary" option will be made visible.
3. From there, the "remove primary" option should be always visible, clearing the value.

It seems that such feature will most likely be a (free) Extension. Because it requires a whole new layer within the program, and it's not really required for non-blog websites.

I'll get back on this as soon as I know more on its implementation.

Cheers!

**Comment by tonybahama** [comment]
1. Yes, that is correct
2. Partially, and only if more than 1 categories are attached to the post/product. If only one is selected, "Primary Category" text is hidden. If you select two categories then the "Make Primary" is selectable next to every attached category.
3. Remove primary is not necessary, only "Make primary" is. If you want to change it, then you "Make Primary" another one. After all there has to be a category in the link.

I made a screenshot of how Yoast solved it:
http://tinypic.com/r/1z1zeol/9

**Comment by sybrew** [comment]
About 3, having no primary would allow it to be automated again.

The list I made is how I'm going to approach this issue 👍. You're spot on about 2.

**Comment by tonybahama** [comment]
About 3: Yup, that makes sense as well. I would like to point out that when you create a new product/post, attach 2 categories it would still be automated. Only the "possibility" shows up. 

I more thing. This should be saved as a meta in the post, so importing would be possible with WP All Import for example by setting _primary_category to category ID "14".

I like your approach and helpfulness, I hope we (or someone) can work out something sooner than 6 months. We can't wait to switch from yoast.

I would code it if I had the knowledge but I am not familiar with code, only logic :)

**Comment by sybrew** [comment]
It's planned to be implemented as late as 2.9.0 (Jan~Mar), and as early as 2.8.0 (December).

Code progress will be pasted against this topic.

**Comment by tonybahama** [comment]
Awesome news! Let me know if you have questions.
Keep up the good work!

P.S.
Thanks, I appreciate it it 😄  I'll stay with logic as an entrepreneur though, but I like to understand what is going on behind the scenes. It's funny because a freelancer dev I'm working with just wanted to hire me as CFO if my business goes bankrupt... 😆

**Comment by entr** [comment]
In the meanwhile can we make the breadcrumb trees filterable in `AutoDescription_Generate_Ldjson::ld_json_breadcrumbs_post()` and `AutoDescription_Generate_Ldjson::ld_json_breadcrumbs_page()`. That would be quite useful too.

**Comment by sybrew** [comment]
Hi Zlatko,

The following filter will be added in 2.7.1:
 `the_seo_framework_breadcrumb_post_sorting_callback`

See https://github.com/sybrew/the-seo-framework/commit/de70ccda1ace1ea734a9ae768920767466c3534a.

Because the trees are being filtered by a PHP "language constructor" (or something in the likes of), you're aught to supply a callback function, from there, the parameter `$trees` will be fed.
You can sort the trees there, be sure to return them.

Example:
```
add_filter( 'the_seo_framework_breadcrumb_post_sorting_callback', 'my_callback_initiator', 10, 2 );
function my_callback_initiator( $func = '', $trees = array() ) {
	return 'my_trees_filter_function';
}

function my_trees_filter_function( $trees = array() ) {

	/* Sort $trees here */

	/* You could also remove entries here. You really shouldn't. Use the WordPress core filters instead. */

	return $trees;
}
```

Cheers!

**Comment by entr** [comment]
Thank you for your quick response. Can you elaborate a bit about that comment

> You could also remove entries here. You really shouldn't. Use the WordPress core filters instead.

Not sure which core filters you're referring to. Filtering `get_the_terms()` seems quite difficult to target only the calls that originate from `ld_json_breadcrumbs_post()` I think.

**Comment by sybrew** [comment]
Hi @entr,

Something like:

```
add_action( 'the_seo_framework_do_before_output', 'my_filter_used_cats' );
/**
 * Filters cats JIT.
 * @see https://github.com/sybrew/the-seo-framework/blob/de70ccda1ace1ea734a9ae768920767466c3534a/inc/classes/generate-ldjson.class.php#L347
 * @see https://developer.wordpress.org/reference/functions/get_the_terms/
 */
function my_filter_used_cats() {

	/* Adjust WordPress Core filter */
	add_filter( 'get_the_terms', 'my_terms_filter', 10, 3 );

}

function my_terms_filter( $terms, $post_id, $taxonomy ) {

	/* filter terms here */

	return $terms;
}

add_action( 'the_seo_framework_do_after_output', 'remove_my_terms_filter' );
/**
 * Removes terms filter, so the rest of the website acts normally.
 */
function remove_my_terms_filter() {
	remove_filter( 'get_the_terms', 'my_terms_filter', 10 );
}
```

But rethinking it, this seems to be a lot more work than just filtering the terms with the provided sorting filter...
The reason I pointed you in this direction is that the filter is truly meant for sorting, and any code changes might not consider removal of the trees.

I'll add another filter in a moment which is more suitable for filtering the terms in the breadcrumbs. Cheers 😄

**Comment by entr** [comment]
@sybrew Just perfect. Looking forward to next release.

**Comment by tonybahama** [comment]
Hi!
It's been a while. Is this feature going to be included in 2.9 then? :)

**Comment by sybrew** [comment]
There's a lot planned for 2.9, so I might need to shift some things to 3.0.
The most important thing for 2.9 is to add a tabular interface to inpost SEO settings; if that's easy to do then this will see the light of day then too.

**Comment by tonybahama** [comment]
Ah okay. I can't wait to switch to your plugin for 1 year now but without this feature it would break our site :/

I really hope you can make it work :)

On 2017. Feb 21., Tue at 17:51 Sybre Waaijer

<
mailto:Sybre Waaijer <notifications@github.com>
> wrote:

a, pre, code, a:link, body { word-wrap: break-word !important; }

There's a lot planned for 2.9, so I might need to shift some things to 3.0.

The most important thing for 2.9 is to add a tabular interface to inpost SEO settings; if that's easy to do then this will see the light of day then too.

—

You are receiving this because you commented.

Reply to this email directly,
https://github.com/sybrew/the-seo-framework/issues/32#issuecomment-281403940
, or

**Comment by sybrew** [comment]
Moved to 3.0.0 (it's nothing more special than a 2.x or 3.x release), because 2.9.0 will fix [too many too prominently available bugs](https://github.com/sybrew/the-seo-framework/blob/98f360692635da837805e8a076f9b212b077706e/readme.txt#L218-L231).
Thus I'm speeding up the current process.

In short, the primary category selection is still planned with the same ETA as before. Its version number only has changed 😄

**Comment by tonybahama** [comment]
Haha I understand :)

Keep up the good work!

K.Szabó Levente

On 2017. Mar 03., Fri at 19:06 Sybre Waaijer

<
mailto:Sybre Waaijer <notifications@github.com>
> wrote:

a, pre, code, a:link, body { word-wrap: break-word !important; }

Moved to 3.0.0 (it's nothing more special than a 2.x or 3.x release), because 2.9.0 will fix
https://github.com/sybrew/the-seo-framework/blob/98f360692635da837805e8a076f9b212b077706e/readme.txt#L218-L231
.

Thus I'm speeding up the current process.

In short, the primary category selection is still planned with the same ETA as before. Its version number only has changed

😄

—

You are receiving this because you commented.

Reply to this email directly,
https://github.com/sybrew/the-seo-framework/issues/32#issuecomment-284026977
, or

**Comment by sybrew** [comment]
**Feature in progress...**

# What will happen:
- The selection will be added.
- Breadcrumb script amount will be reduced to 1, instead of using multiple. 
   - The primary category tree will be used.
   - Google no longer supports multiple breadcrumbs anyway and will ignore the others.

## Culprit: We want to have no-JS compatibility.

What this means is that the feature should be available for users with and without JavaScript.

Now, there's only one feasible filter (I've found so far) that allows this. 
However, that filter is used for naming the categories; not for implementing options.

The function: `wp_link_category_checklist()`
The filter: `the_category`

Because of the above statements we should steer away from no-JS compatibility.
It will then become more maintainable, and as the web (and WordPress) is moving forward to JS-only (React, Vue, REST, Calypso, etc.), it's only logical to use it to its fullest extent.

Even though some users will falter...

### Working alternative: JS-only.

With a JS-only implementation, no filters will be used. It's much more feasible, and it's not so much of a "hack".

Through JavaScript, we add buttons that will highlight and allow you to enable the primary category.

### Another alternative: Purge the meta box, and replace it.

Within `edit-link-form.php`, the meta box is registered.
We can remove and replace the meta box with our own version.

This can break WordPress' implementation of the meta box. Unexpected behavior can come when WordPress updates the meta box and the TSF version used is lacking that update.

For this reason, we must expulse this resolution.

### Thoughts?

We're going to start on the JS-only alternative. If you have anything to contribute, before it's released, do let us know!

## Culprit: Migration to this new setting

Currently, the primary category is selected through an alphabetic order.

To maintain this behavior, it'll probably be most likely to maintain this for posts that have been set prior to this update.

## Culprit: Default behavior

Before the user selects a primary category, a default one has to be highlighted.
This will probably be the first one checked, i.e. in alphabetic order.

If the user hasn't highlighted the primary category manually, it should automatically be assigned when a higher-up category is enabled.

If the user has highlighted the primary category manually, it should stick.
For this reason, two states need to be used. One is "highlight/probable", and one is definitively "selected".

The just-highlighted category shouldn't be saved in the post meta.

## Culprit: We don't use "better" post meta, yet.

See #185.

This option will be the first entry. Therefore, this feature must set prerequisites for 185's implementation.

# Summing it up

We know what's ahead. And it's quite easily done from my perspective 😄

**Please contribute if you have anything to add.**

This will likely be a three-day project. 
When it's done and released, suggestions are harder to implement and they will take much more time to do so.

**Comment by tonybahama** [comment]
Awesome!
Question: will you save it as a post meta what the primary category id is?

Why? Because for example we refer to this many times right now, also create dependencies. By storing it as a post meta ( like the_seo_framework_primarycat) it would be easy for us to maintain compatibility.

**Comment by sybrew** [comment]
@tonybahama Yes, we are and will still be using WordPress Meta API 😄

**Comment by sybrew** [comment]
FYI, this is how it looks:

![image](https://user-images.githubusercontent.com/7662236/31803394-bcb85bce-b553-11e7-9fb8-b24a7ad1aa4f.png)

I would love feedback! Especially since I'm not 100% happy about it.

**Comment by tonybahama** [comment]
Nice work!

How does it store which data is selected?

Üdv,
K.Szabó Levente

**Comment by sybrew** [comment]
Hi @tonybahama,

That's still to be determined 😄 I'll follow up on that when it's ready.

**Comment by sybrew** [comment]
This will be the layout:

![image](https://user-images.githubusercontent.com/7662236/31844526-7c0caff2-b5f9-11e7-883e-3a0ed20abac2.png)

------------

@tonybahama 

The data is stored through the Post Meta API, as such:
`'_primary_term_' . $taxonomy`

Example:
```php
get_post_meta( $post_id, '_primary_term_' . $taxonomy, true );
````

This method will allow you to fetch set primary term ID, cached within TSF's class object:
```php
/**
 * @param int|null $post_id     The post ID. When empty, it will get the current post.
 * @param string   $taxonomy Required. The taxonomy name.
 * @return int The primary term ID. 0 if not set. 
 */
the_seo_framework()->get_primary_term_id( $post_id, $taxonomy );
```

The method above doesn't guarantee if the term exists. Users can assign a primary term but may delete the term later.
So, this method will test for the term's existence, and will return the term object, cached through WordPress' object cache:
```php
/**
 * @param int|null $post_id The post ID. When empty, it will get the current post.
 * @param string   $taxonomy Required. The taxonomy name.
 * @return \WP_Term|false The primary term. False if not set.
 */
the_seo_framework()->get_primary_term( $post_id, $taxonomy );
```

**Comment by 1ucay** [comment]
If I call get_the_category( $post->ID ), it would be nice to have the first position primary term.
It is possible extended orderby term filter via
https://developer.wordpress.org/reference/hooks/get_terms_orderby/
https://developer.wordpress.org/reference/hooks/terms_clauses/

sybrew - good example is in plugin Anything Order by Terms https://cs.wordpress.org/plugins/anything-order-by-terms/
I can overide term order via change args
`get_terms( array( 'orderby' => 'name term_order'`


In your plugin we can have like this
`get_terms( array( "orderby" => "name term_primary term_order"`

What do you think?

**Comment by sybrew** [comment]
Hi @1ucay, 

I've looked into that when implementing this system, and I believe it'll interfere with things it ultimately shouldn't. WordPress consists of many (often broken) pieces, and this will add to that puzzle.

For example, a plugin might want to fetch a simple list of terms in default order. Then, TSF will go ahead and change that, even when the said plugin doesn't expect it.
Another use case is found in the 2-star review for the linked-to plugin.

In any case, I think it's best to leave this issue as-is and open new ones for features that build onto this when necessary 😄.
